### PR TITLE
Copter: rework rangefinder state to estimate terrain height

### PR DESCRIPTION
**Pull request by jschall**
This has the advantage of interpolating the resulting height between rangefinder measurements and eliminates lag introduced by the terrain height filter.

This has so far only been tested in SITL.

### CODE REVIEW

File: ArduCopter/sensors.cpp [hunk 1]
```diff
@@ -44,44 +44,62 @@ void Copter::read_rangefinder(void)
...
+
+bool Copter::get_rangefinder_height_above_terrain(int32_t& ret)
+{
+    int32_t height_above_origin_cm;
+
+    if (!rangefinder_state.enabled || !rangefinder_state.terrain_height_healthy || !current_loc.get_alt_cm(Location_Class::ALT_FRAME_ABOVE_ORIGIN, height_above_origin_cm)) {
+        return false;
+    }
+
+    ret = MAX(0, height_above_origin_cm-rangefinder_state.terrain_height_filt_cm.get());
```
Comment (derekxm) [inline ArduCopter/sensors.cpp]: What's the logic here? is hight_above_origin_cm the relative altitude?

**Comment by jschall** [comment]
Flight tested. Works as intended, except that the filter is too slow. Here's a plot of it stepping on and off of a 1.5m tall landing pad during a bad precision landing:
figure_1

File: ArduCopter/sensors.cpp (hunk 1 above)
Comment (jschall) [inline ArduCopter/sensors.cpp]: * ```height_above_origin_cm``` is the vehicle's height above the EKF's origin altitude.
* ```terrain_height_filt_cm.get()``` is the height of terrain above the EKF's origin altitude.

Subtracting the terrain's height above origin from the vehicle's height above origin results in the vehicle's height above terrain.

**Comment by jschall** [comment]
I'm going to test with the terrain filter set to 3hz instead of .25hz. If needed we can have filtered and unfiltered terrain height.

**Comment by OXINARF** [comment]
@jschall Still working on this or can it be closed? It's more than 2 years old.

**Pull request closed without merging**